src/main/java/htsjdk/samtools/cram/structure/Container.java

```diff
@@ -67,11 +67,28 @@
     // slices found in the container:
     public Slice[] slices;
 
-    // for indexing:
+    // this Container's byte offset from the the start of the stream.
+    // Used for indexing.
+    private long byteOffset;
+
+    public long getByteOffset() {
+        return byteOffset;
+    }
+
     /**
-     * Container start in the stream, in bytes.
+     * Set this Container's byte offset from the the start of the stream.
+     * Also distribute this value to the Container's constituent {@link Slice}s.
+     * For indexing.
+     * @param byteOffset the byte location in the stream where this Container begins
      */
-    public long offset;
+    public void setByteOffset(final long byteOffset) {
+        this.byteOffset = byteOffset;
+        if (slices != null) {
+            for (final Slice slice : slices) {
+                slice.containerByteOffset = byteOffset;
+            }
+        }
```

Comment:
@jmthibault79 It seems like it should be possible to reduce the number of call sites that need to update these offsets. Every call to this method (except for the one in ContainerStreamWriter, when a new container is being created from records) is immediately preceded by a call (either directly or indirectly), to ContainerIO.readContainer. It would be much  simpler to make ContainerIO responsible for all of those cases, even if it means passing the offset in (although I think it could often derive the offset from the stream position). Would that work ? I think it would reduce the complexity around understanding who/when the slice offsets get propagated, etc.

Reply:
I do think this is possible, and I have started an attempt at this.  I'll work on that some more.

Good point about propagation.  I will have to think more about that.

Reply:
Addressed both of these in a new commit.